src/validations/rules/ssp.sch

```diff
@@ -2209,38 +2209,61 @@
                 doc:template-reference="System Security Plan Template §15 Attachment 13"
-                id="inventory-item-has-ipv4-and-ipv6"
+                id="inventory-item-network-address"
                 role="error"
-                test="o:prop[@name eq 'ipv4-address'] and o:prop[@name eq 'ipv6-address']">An inventory item must have an ipv4 address and an ipv6 address.</sch:assert>
+                test="if (o:prop[@name eq 'ipv4-address'])
+                then (o:prop[@name eq 'ipv6-address'])
+                else (true())">If any inventory-item has a prop with a name of 'ipv4-address' it must also have a prop with a name of 'ipv6-address'</sch:assert>
             <sch:assert
                 diagnostics="ipv4-has-content-diagnostic"
                 doc:guide-reference="Guide to OSCAL-based FedRAMP System Security Plans §6.5.1 and §6.5.2"
                 doc:template-reference="System Security Plan Template §15 Attachment 13"
                 id="ipv4-has-content"
                 role="error"
-                test="o:prop[@value ne '0.0.0.0' and normalize-space(@value) ne '' and @name eq 'ipv4-address']">
-                <xsl:value-of select="o:prop[@name='asset-id']/@value"/> does not have an appropriate IPv4 value.</sch:assert>
+                test="if(o:prop[@name eq 'ipv4-address'])
+                then(o:prop[matches(@value, '(^[0-9][0-9]?[0-9]?\.[0-9][0-9]?[0-9]?\.[0-9][0-9]?[0-9]?\.[0-9][0-9]?[0-9]?$)')])
+                else(true())"><xsl:value-of select="o:prop[@name='asset-id']/@value"/> must have an appropriate IPv4 value.</sch:assert>
+            <sch:assert
+                diagnostics="ipv4-has-non-placeholder-diagnostic"
+                doc:guide-reference="Guide to OSCAL-based FedRAMP System Security Plans §6.5.1 and §6.5.2"
+                doc:template-reference="System Security Plan Template §15 Attachment 13"
+                id="ipv4-has-non-placeholder"
+                role="error"
+                test="if(o:prop[@name eq 'ipv4-address'])
+                then(o:prop[matches(@value, '0.0.0.0')])
+                else(false())"><xsl:value-of select="o:prop[@name='asset-id']/@value"/> must have an appropriate IPv4 value.</sch:assert>
             <sch:assert
                 diagnostics="ipv6-has-content-diagnostic"
                 doc:guide-reference="Guide to OSCAL-based FedRAMP System Security Plans §6.5.1 and §6.5.2"
                 doc:template-reference="System Security Plan Template §15 Attachment 13"
                 id="ipv6-has-content"
                 role="error"
-                test="o:prop[@value ne '0000:0000:0000:0000' and normalize-space(@value) ne '' and @name eq 'ipv6-address']">
-                <xsl:value-of select="o:prop[@name='asset-id']/@value"/> does not have an appropriate IPv6 value.</sch:assert>
+                test="if(o:prop[@name eq 'ipv6-address'])
+                then(o:prop[matches(@value, '([A-Fa-f0-9]{1,4}:){7}[A-Fa-f0-9]{1,4}')])
```

Comment:
This regex does not handle typically encountered "compressed" IPv6 addresses (e.g., `2604:a880:0:1010::655:8001` and `::1` and `::` - that latter one is one which should be rejected, not by the regex, but by an additional assertion).

The best IPv6 address regex I have found thus far  is
`(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))`.

The best regex tester I have found is https://regex101.com/.

Reply:
I tried that one already I believe if it is the one you suggested elsewhere.  I had schematron errors on it and couldn't isolate why.  I'll try again and provide you with the error.

The problem likely is XPath regex vs 'regular' regex.  But the above is so dense as to be unreadable so we will have to see how far we can get.

Reply:
The XPath expression `matches ('::','(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))` is successfully evaluated (and returns `true`) within a contemporary copy of OxygenXML when the evaluation is performed at the XPath 3.1 and XPath 2.0 levels. That signifies that the XPath expression, and the regular expression, and the evaluation of the XPath expression are all valid at the 3.1 and 2.0 level.

Is the error attributable to the XSpec implementation used? If so, it would be appropriate to separately determine if an in-OxygenXML Schematron evaluation (as opposed to outboard XSpec) succeeds or fails (OxygenXML uses a custom Schematron evaluation for inline schema-driven editing)? We have previously observed unusual and unexpected problems with XSpec implementations and this might be an opportunity to determine the nature of and avoidance of such problems.

The error you mention has an air of improper text value (as in expand-text='true') template handling though that may not have anything to do with the problem.

Reply:
The move to a variable does solve the problem.  Good catch on the text expand likely being the issue.  I've made the change and the code now runs smoothly.